Stop generating autograd functions for derivatives.yaml entries that only specify output differentiability.

gchanan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #19431 Get rid of unnecessary matches_jit_signature: True specifications.
* #19430 Make one_hot non-differentiable.
* #19429 Remove 'BoolTensor', 'IndexTensor' from frontend specifications.
* #19428 Have _embedding_bag_dense_backward match JIT signature.
* #19427 Have embedding_dense_backward match JIT signature.
* #19426 Hook up non_differentiability in derivatives.yaml when no autograd function is generated.
* #19425 Move non_differentiable_arg_names from autograd functions to differentiability_info.
* **#19424 Stop generating autograd functions for derivatives.yaml entries that only specify output differentiability.**
* #19272 Rename 'not_differentiable' to 'non_differentiable'.

the codegen for derivatives puts information about what is differentiable (inputs and outputs) together with the autograd function.
This is not a good idea, because it ties together the concept of "having an autograd function" with "specifying differentiability".

This results in a number of hacks.  One we fix here is wrt 'output_differentiability'.
Previously, we would generate autograd functions for entires that only specify output differentiability, e.g. values()/_values(),
and then warn that the actual autograd function was ignored.

Now, we avoid generating the autograd function and error out if we ever ignore an actual derivative specification.

Generated code difference: https://gist.github.com/gchanan/e10df0c68f9f9ca4dee33458763a24b7.

Differential Revision: [D15003380](https://our.internmc.facebook.com/intern/diff/D15003380)

gchanan committed:
Stop generating autograd functions for derivatives.yaml entries that only specify output differentiability.

the codegen for derivatives puts information about what is differentiable (inputs and outputs) together with the autograd function.
This is not a good idea, because it ties together the concept of "having an autograd function" with "specifying differentiability".

This results in a number of hacks.  One we fix here is wrt 'output_differentiability'.
Previously, we would generate autograd functions for entires that only specify output differentiability, e.g. values()/_values(),
and then warn that the actual autograd function was ignored.

Now, we avoid generating the autograd function and error out if we ever ignore an actual derivative specification.

Generated code difference: https://gist.github.com/gchanan/e10df0c68f9f9ca4dee33458763a24b7.

cpuhrsch commented on tools/autograd/load_derivatives.py:
```diff
@@ -207,8 +217,14 @@ def unzip(xs):
-                                    signature, canonical, output_differentiability)
+    autograd_fn = None
+
+    # only create an autograd function if we are actually going to calculate a derivative
```
Does this mean 'only create an autograd function, if we actually plan to use it'?

cpuhrsch replied:
Or "don't create an autograd function, if there is no derivative available", which means it can't possibly be differentiable.

gchanan replied:
It means the former, although I prefer the wording in the diff.

The later i slightly wrong because if no derivatives are specified it can still fall through, i.e. be differentiable.

cpuhrsch replied:
Well, we can calculate a derivative and have an implementation, but not provide it as part of autograd or plan to do so.

The term "derivative" here is confusing, because there's operators with explicit derivative implementations (i.e. xyz_backward) and some operators where the derivative is a result of the fact that they're implemented as a string of already differentiable operators. 

It'd be nice to have a clear distinction between those cases.

cpuhrsch commented on tools/autograd/load_derivatives.py:
```diff
@@ -19,18 +19,30 @@ def load_derivatives(path, declarations):
-def create_autograd_function(name, derivatives, args_with_derivatives, non_differentiable_arg_names,
-                             signature, declaration, output_differentiability):
+def create_autograd_function(name, derivatives, args_with_derivatives,
+                             declaration, output_differentiability,
```
Is the 'output_differentiability' argument still needed?

gchanan replied:
No, good catch.

gchanan committed:
Update on "Stop generating autograd functions for derivatives.yaml entries that only specify output differentiability."

Stop generating autograd functions for derivatives.yaml entries that only specify output differentiability.

the codegen for derivatives puts information about what is differentiable (inputs and outputs) together with the autograd function.
This is not a good idea, because it ties together the concept of "having an autograd function" with "specifying differentiability".

This results in a number of hacks.  One we fix here is wrt 'output_differentiability'.
Previously, we would generate autograd functions for entires that only specify output differentiability, e.g. values()/_values(),
and then warn that the actual autograd function was ignored.

Now, we avoid generating the autograd function and error out if we ever ignore an actual derivative specification.

Generated code difference: https://gist.github.com/gchanan/e10df0c68f9f9ca4dee33458763a24b7.

gh-metadata: pytorch pytorch 19424 gh/gchanan/17/head

gchanan committed:
Update on "Stop generating autograd functions for derivatives.yaml entries that only specify output differentiability."

Stop generating autograd functions for derivatives.yaml entries that only specify output differentiability.

the codegen for derivatives puts information about what is differentiable (inputs and outputs) together with the autograd function.
This is not a good idea, because it ties together the concept of "having an autograd function" with "specifying differentiability".

This results in a number of hacks.  One we fix here is wrt 'output_differentiability'.
Previously, we would generate autograd functions for entires that only specify output differentiability, e.g. values()/_values(),
and then warn that the actual autograd function was ignored.

Now, we avoid generating the autograd function and error out if we ever ignore an actual derivative specification.

Generated code difference: https://gist.github.com/gchanan/e10df0c68f9f9ca4dee33458763a24b7.

gh-metadata: pytorch pytorch 19424 gh/gchanan/17/head
